community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/StringSchemaKey.java

```diff
@@ -149,24 +161,57 @@ private static int byteArrayCompare( byte[] a, byte[] b, boolean ignoreLength )
             }
         }
 
-        return ignoreLength ? 0 : Integer.compare( a.length, b.length );
+        return ignoreLength ? 0 : Integer.compare( aLength, bLength );
     }
 
     @Override
     public String toString()
     {
-        return format( "value=%s,entityId=%d,bytes=%s", asValue(), getEntityId(), Arrays.toString( bytes ) );
+        return format( "value=%s,entityId=%d,bytes=%s",
+                asValue(),
+                getEntityId(),
+                bytes == null ? "null" : Arrays.toString( Arrays.copyOf( bytes, bytesLength ) ) );
     }
 
     @Override
     public void writeString( String value )
     {
         bytes = UTF8.encode( value );
+        bytesLength = bytes.length;
+        bytesDereferenced = false;
     }
 
     @Override
     public void writeString( char value )
     {
         writeString( String.valueOf( value ) );
     }
+
+    void copyFrom( StringSchemaKey key )
+    {
+        setBytesLength( key.bytesLength );
+        System.arraycopy( key.bytes, 0, bytes, 0, key.bytesLength );
+        setEntityId( key.getEntityId() );
+        setCompareId( key.getCompareId() );
+    }
+
+    /**
+     * Ensures that the internal byte[] is long enough, or longer than the given {@code length}.
+     * Also sets the internal {@code bytesLength} field to the given {@code length} so that interactions with the byte[]
+     * from this point on will use that for length, instead of the length of the byte[].
+     *
+     * @param length minimum length that the internal byte[] needs to be.
+     */
+    void setBytesLength( int length )
+    {
+        if ( bytesDereferenced || bytes == null || bytes.length < length )
+        {
+            bytesDereferenced = false;
+
+            // allocate a bit more than required so that there's a higher chance that this byte[] instance
+            // can be used for more keys than just this one
+            bytes = new byte[length + length / 2];
```

Comment:
do you think it makes sense to have something that will trim bytes eventually if they will grow over time because of some big value?

Reply:
Hmm, well... that's a bit more complex. These key instances only lives during the result "iterator" anyway, so usually not an issue.

Reply:
ok